File: lib/encounter/tier_mismatch.rb
module Encounter
  # Tier Mismatch modifiers (encounter_design.md -> Tier Mismatch).
  #
  # When two Creatures of different Tiers oppose each other, the
  # higher-Tier Creature gains advantages scaled by the Tier difference
  # `delta = higher.tier - lower.tier`. Tier 0 counts as 0.5 per project
  # convention; magnitudes are floored to integers.
  #
  #   - Inherent damage reduction: 5 * delta against a lower-Tier
  #     attacker (an Inherent Bonus).
  #   - Ascendancy: +/- 2 * delta on opposed checks (Bonus for the
  #     higher Creature, Penalty for the lower).
  #
  # An ability may raise a Creature's *effective* Tier for a single
  # resolution (the Glory Channel Divinity Glorious Charge adds +1);
  # callers fold that into the tier they pass.
  #
  # This Ruby module owns the *Inherent damage reduction* half, applied
  # server-side when damage is resolved. The *Ascendancy* check modifier
  # is computed client-side in the JS Check Resolution engine
  # (public/js/tierMismatch.js), since Rolls resolve there.
  module TierMismatch
    module_function

    # Tier 0 -> 0.5 per project convention.
    def effective_tier(tier)
      t = tier.to_f
      t.zero? ? 0.5 : t
    end

    # Signed Tier difference actor - opponent (using the 0 -> 0.5 rule).
    def delta(actor_tier, opponent_tier)
      effective_tier(actor_tier) - effective_tier(opponent_tier)
    end

    # Inherent damage reduction the defender gets against a lower-Tier
    # attacker's hit: `Inherent Damage Reduction Per Tier` (encounter_config.yaml,
    # default 5) per Tier the defender is above the attacker, else 0.
    def inherent_damage_reduction(defender_tier, attacker_tier)
      d = delta(defender_tier, attacker_tier)
      return 0 unless d.positive?
      (Config.inherent_damage_reduction_per_tier * d).floor
    end
  end
end

File: public/js/tierMismatch.js
// Tier Mismatch (encounter_design.md -> Tier Mismatch).
//
// When two Creatures of different Tiers oppose each other on a Check, the
// higher-Tier side gains an Ascendancy Bonus and the lower-Tier side takes
// an Ascendancy Penalty, scaled by the Tier difference (+/- 2 per Tier).
// Tier 0 counts as 0.5; magnitudes are floored.
//
// This is the Check-side half of the rule. The Inherent damage reduction
// half is not a Roll modifier — it reduces dealt damage — and is applied
// server-side when damage is resolved.
//
// Each Roll may carry a numeric `tier`. apply() runs AFTER cross-side
// Propagation and appends the Ascendancy entry directly from the two
// sides' tiers, so it is not itself inverted/propagated. A Roll without a
// tier (or with no opposing tier) gets nothing.
export class TierMismatch {
  static ASCENDANCY_PER_TIER = 2;
  static TYPE = 'Ascendancy';

  // Tier 0 -> 0.5 per project convention. An absent tier (null/undefined)
  // returns null — distinct from Tier 0 — so a Roll with no Tier opts out.
  static effectiveTier(tier) {
    if (tier === null || tier === undefined) return null;
    const t = Number(tier);
    if (!Number.isFinite(t)) return null;
    return t === 0 ? 0.5 : t;
  }

  // The Ascendancy [type, amount] pair for an actor of `actorTier` opposing
  // `opponentTier`, or null when either tier is absent, the tiers are equal,
  // or the floored magnitude is 0.
  static ascendancyModifier(actorTier, opponentTier) {
    const a = TierMismatch.effectiveTier(actorTier);
    const o = TierMismatch.effectiveTier(opponentTier);
    if (a === null || o === null) return null;
    const delta = a - o;
    if (delta === 0) return null;
    const magnitude = Math.floor(TierMismatch.ASCENDANCY_PER_TIER * Math.abs(delta));
    if (magnitude === 0) return null;
    return [TierMismatch.TYPE, delta < 0 ? -magnitude : magnitude];
  }

  // Append each Roll's Ascendancy entry, measured against the opposing
  // side's primary (initiating / defending) Roll's tier — mirroring the
  // pairing used by Propagation. Returns a copy; input is not mutated.
  static apply(check) {
    const supporting = check.supporting || [];
    const opposing = check.opposing || [];
    const initiating = supporting[0] || null;
    const defending = opposing[0] || null;

    const extend = (roll, opponent) => {
      if (!roll) return roll;
      const mod = opponent ? TierMismatch.ascendancyModifier(roll.tier, opponent.tier) : null;
      if (!mod) return roll;
      return { ...roll, bonusPenaltyList: (roll.bonusPenaltyList || []).concat([mod]) };
    };

    return {
      supporting: supporting.map((roll) => extend(roll, defending)),
      opposing: opposing.map((roll) => extend(roll, initiating)),
    };
  }
}
